napari/utils/events/evented_model.py

```diff
@@ -55,15 +71,112 @@ def no_class_attributes():
+    - https://github.com/samuelcolvin/pydantic/issues/4161
+    """
+
+    def _type_analysis(self):
+        super()._type_analysis()
+        # if class validators are unset, it means pydantic ignored them, so
+        # we add them back in just like pydantic does
+        if not self.class_validators:
+            get_validators = getattr(
+                self.outer_type_, '__get_validators__', dict
+            )
+            self.class_validators.update(
+                {
+                    f'{self.outer_type_.__name__}_{i}': Validator(
+                        validator, pre=True
+                    )
+                    for i, validator in enumerate(get_validators())
+                }
+            )
+
+        origin = get_origin(self.outer_type_)
+        # since issubclass(collections.abc.Set, typing.Set) == False, and our EventedSet
+        # is a collections.abc.Set, we have to tell pydantic to treat it the same as the other sets
+        # isinstance(origin, type) is needed for non-class origins (e.g: Literal)
+        if (
+            origin is not None
+            and isinstance(origin, type)
+            and issubclass(origin, abc.Set)
+        ):
+            self.shape = SHAPE_SET
+            self.type_ = get_args(self.outer_type_)[0]
+
+    def populate_validators(self):
+        super().populate_validators()
+        # we need to get rid of the basic arbitrary_types validator because otherwise
+        # in cases where we don't coerce type (e.g: Colormap) it always fails cause the type is wrong
+        # We then coerce the type ourselves below
+        self.validators = [
+            f
+            for f in self.validators
+            if f.__name__ != 'arbitrary_type_validator'
+        ]
+
+        origin = get_origin(self.outer_type_) or self.outer_type_
+        # anything that's not really a class should not be used for coercion
+        if (
+            not isinstance(origin, type)
+            or origin is Union
+            or issubclass(origin, Protocol)
+        ):
+            return
+
+        # we should get here only if origin is a *real* class and we want to coerce it
+        def coerce_type(cls, value, field):
+            """
+            Last validator to be run, ensures the returned type matches the annotation.
+            """
+            if field.allow_none and value is None:
+                return None
+            if isinstance(value, origin):
+                return value
+            return origin(value)
+
+        self.post_validators = self.post_validators or []
+        self.post_validators.extend(prep_validators([coerce_type]))
+
+
+@contextmanager
+def parametrized_generic_fix():
+    """
+    Context in which pydantic more consistently validates parametrized generics.
+
+    This context temporarily replaces ModelField with
+    ParametrizedGenericCompliantModelField.
+    It also permanently overrides pydantic.sequence_like() to be more lax.
```

Comment:
This is an intentional side effect: without permanently replacing `pydantic`'s `sequence_like()` with a more lenient version, many types will simply fail when parametrized (see the end of [this issue in pydantic](https://github.com/samuelcolvin/pydantic/issues/4161) for an example)